app/models/conversion_host.rb

```diff
@@ -124,28 +208,61 @@ def resource_supports_conversion_host
     end
   end
 
-  def check_resource_credentials(fatal = false, extra_msg = nil)
-    success = send("check_resource_credentials_#{resource.ext_management_system.emstype}")
-    if !success and fatal
-      msg = "Credential not found for #{resource.name}."
-      msg += " #{extra_msg}" unless extra_msg.blank?
-      _log.error(:msg)
+  # Find the associated credentials for the current conversion host based
+  # on the type of provider. Each provider should look for a directly
+  # associated authentication first, and default to the authentication
+  # for the resource second.
+  #--
+  # TODO: Each provider subclass should define their own method for
+  # finding credentials.
+  #
+  def find_credentials(msg = nil)
+    authentication = send("find_resource_credentials_#{resource.ext_management_system.emstype}")
+
+    unless authentication
+      msg = "Credentials not found for conversion host #{name} or resource #{resource.name}"
+      msg << " #{msg}" if msg
+      _log.error(msg)
       raise MiqException::Error, msg
     end
-    success
+
+    authentication
   end
 
-  def check_resource_credentials_rhevm
-    !(resource.authentication_userid.nil? || resource.authentication_password.nil?)
+  # Find the credentials for the associated Redhat resource. If an AuthUseridPassword
+  # object is not found, but the associated userid and password are found, return
+  # a (temporary, unsaved) AuthUseridPassword object using those values.
+  #--
+  # TODO: Move this into the provider specific subclass.
+  #
+  def find_credentials_redhat
+    authentication = AuthUseridPassword.find_by(:resource => resource)
+
+    if authentication.nil? && resource.authentication_userid && resource.authentication_password
+      authentication = AuthUseridPassword.new(
+        :name     => resource.name,
+        :resource => resource,
+        :userid   => resource.authentication_userid,
+        :password => resource.authentication.password
+      )
+    end
+
+    authentication
   end
 
-  def check_resource_credentials_openstack
-    ssh_authentications = resource.ext_management_system.authentications
-                                  .where(:authtype => 'ssh_keypair')
-                                  .where.not(:userid => nil, :auth_key => nil)
-    !ssh_authentications.empty?
+  # Find the credentials for the associated with the Openstack resource. This will
+  # look for an AuthPrivateKey associated with the resource, or the EMS of the
+  # associated resource if one cannot be found.
+  #--
+  # TODO: Move this into the provider specific subclass.
+  #
+  def find_credentials_openstack
+    AuthPrivateKey.find_by(:resource => resource) || resource.ext_management_system.authentication_type('ssh_keypair')
```

Comment:
Would be better to use the `resource.authentication_*` accessors instead of doing direct find_by as this ignores the auth type.

Reply:
Ok, updated.